src/eh1/io.rs

```diff
@@ -0,0 +1,684 @@
+            expected_err: None,
+        }
+    }
+
+    /// Create a read ready transaction
+    pub fn read_ready(ready: bool) -> Transaction {
+        Transaction {
+            expected_mode: Mode::ReadReady(ready),
+            expected_data: Vec::new(),
+            response: Vec::new(),
+            expected_err: None,
+        }
+    }
+
+    /// Create a fill buffer transaction
+    pub fn fill_buf(response: Vec<u8>) -> Transaction {
+        Transaction {
+            expected_mode: Mode::FillBuff,
+            expected_data: Vec::new(),
+            response,
+            expected_err: None,
+        }
+    }
+
+    /// Create a consume transaction
+    pub fn consume(consumed: usize) -> Transaction {
+        Transaction {
+            expected_mode: Mode::Consume(consumed),
+            expected_data: Vec::new(),
+            response: Vec::new(),
+            expected_err: None,
+        }
+    }
+
+    /// Add an error return to a transaction
+    ///
+    /// This is used to mock failure behaviors.
+    ///
+    /// Note: When attaching this to a read transaction, the response in the
+    /// expectation will not actually be written to the buffer.
+    pub fn with_error(mut self, error: ErrorKind) -> Self {
+        self.expected_err = Some(error);
+        self
+    }
+}
+
+/// Mock IO implementation
+pub type Mock = Generic<Transaction, Vec<u8>>;
+
+impl ErrorType for Mock {
+    type Error = ErrorKind;
+}
+
+impl Write for Mock {
+    fn write(&mut self, buffer: &[u8]) -> Result<usize, Self::Error> {
+        let transaction = self.next().expect("no expectation for io::write call");
+        assert_eq!(
+            transaction.expected_mode,
+            Mode::Write,
+            "io::write unexpected mode"
+        );
+        assert_eq!(
+            &transaction.expected_data, &buffer,
+            "io::write data does not match expectation"
+        );
+
+        match transaction.expected_err {
+            Some(err) => Err(err),
+            None => Ok(buffer.len()),
+        }
+    }
+
+    fn flush(&mut self) -> Result<(), Self::Error> {
+        let transaction = self.next().expect("no expectation for io::flush call");
+        assert_eq!(
+            transaction.expected_mode,
+            Mode::Flush,
+            "io::flush unexpected mode"
+        );
+
+        match transaction.expected_err {
+            Some(err) => Err(err),
+            None => Ok(()),
+        }
+    }
+}
+
+impl Read for Mock {
+    fn read(&mut self, buffer: &mut [u8]) -> Result<usize, Self::Error> {
+        let transaction = self.next().expect("no expectation for io::read call");
+        assert_eq!(
+            transaction.expected_mode,
+            Mode::Read,
+            "io::read unexpected mode"
+        );
+
+        if transaction.response.len() > buffer.len() {
+            panic!("response longer than read buffer for io::read");
+        }
+
+        let len = std::cmp::min(buffer.len(), transaction.response.len());
```

Comment:
Since we already assert above, that the response length can't be larger than the buffer length, can't we just use the response length here?

```suggestion
        let len = transaction.response.len();
```